File: server/shared/dotnet/resilience/Pipeline/ResilientPipeline.cs
// -----------------------------------------------------------------------
// <copyright file="ResilientPipeline.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Resilience.Pipeline;

using D2.Shared.Resilience.CircuitBreaker;
using D2.Shared.Resilience.Retry;
using D2.Shared.Result;

/// <summary>
/// Composes <see cref="IResilientLayer{TKey, TValue}"/> layers into a single
/// callable surface. Each call to <see cref="ExecuteAsync"/> walks the
/// configured layers in outer-first order and converts every terminating
/// exception into a <see cref="D2Result{TValue}"/>, so callers never need
/// try/catch blocks at the call site.
/// </summary>
/// <typeparam name="TKey">
/// Per-call key type (used by Singleflight; ignored by other layers).
/// </typeparam>
/// <typeparam name="TValue">The value type produced by the operation.</typeparam>
/// <remarks>
/// <para>
/// <b>Order matters.</b> The order layers are passed to the constructor IS
/// the protection semantic. The two canonical full-stack compositions:
/// </para>
/// <list type="bullet">
///   <item><description>
///     <c>[Singleflight, CircuitBreaker, Retry]</c> — retry INSIDE the
///     breaker. CB sees one execution per full retry budget. Protects the
///     upstream from sustained pressure (good for fragile upstreams where
///     backoff between attempts gives them air).
///   </description></item>
///   <item><description>
///     <c>[Singleflight, Retry, CircuitBreaker]</c> — retry OUTSIDE the
///     breaker. Each retry is a separate CB execution; when CB opens, the
///     <see cref="CircuitOpenException"/> is treated as transient by the
///     default classifier so the retry layer backs off through it. If the
///     retry budget spans the breaker's cooldown, a later attempt finds the
///     breaker probing / closed and succeeds. Recovers from upstream
///     restarts at the cost of additional caller-side latency. Callers MUST
///     size <see cref="RetryOptions{TValue}.MaxAttempts"/> + backoff to
///     span <see cref="CircuitBreakerOptions.CooldownDuration"/>; otherwise
///     retries exhaust on perpetual <see cref="CircuitOpenException"/> and
///     the pipeline returns
///     <see cref="D2Result{TValue}.ServiceUnavailable"/>.
///   </description></item>
/// </list>
/// <para>
/// <b>Exception → result mapping:</b>
/// </para>
/// <list type="bullet">
///   <item><description>
///     <see cref="CircuitOpenException"/> →
///     <see cref="D2Result{TValue}.ServiceUnavailable"/>
///   </description></item>
///   <item><description>
///     <see cref="OperationCanceledException"/> when the supplied token
///     was the source → <see cref="D2Result{TValue}.Cancelled"/>
///   </description></item>
///   <item><description>
///     Any other exception classified as transient by
///     <see cref="RetryHelper.IsTransientException"/> →
///     <see cref="D2Result{TValue}.ServiceUnavailable"/>
///     (covers transient errors that slipped past the configured layers,
///     e.g. when no Retry layer is configured)
///   </description></item>
///   <item><description>
///     Anything else →
///     <see cref="D2Result{TValue}.UnhandledException"/>
///   </description></item>
/// </list>
/// </remarks>
public sealed class ResilientPipeline<TKey, TValue>
    where TKey : notnull
{
    private readonly IResilientLayer<TKey, TValue>[] r_layersOuterFirst;

    /// <summary>
    /// Initializes a <see cref="ResilientPipeline{TKey, TValue}"/> with the
    /// supplied layers in outer-first order. Layer at index 0 wraps every
    /// subsequent layer; the last layer wraps the operation directly.
    /// Passing zero layers yields a pass-through pipeline that still
    /// performs exception → result mapping.
    /// </summary>
    /// <param name="layersOuterFirst">
    /// Layers in outer-first execution order (see type-level remarks for the
    /// canonical compositions).
    /// </param>
    public ResilientPipeline(params IResilientLayer<TKey, TValue>[] layersOuterFirst)
        => r_layersOuterFirst = layersOuterFirst;

    /// <summary>
    /// Executes <paramref name="operation"/> through the configured layer
    /// stack. Never throws — every terminating exception is converted to a
    /// <see cref="D2Result{TValue}"/> per the type-level mapping.
    /// </summary>
    /// <param name="key">Per-call key (used by Singleflight, ignored by other layers).</param>
    /// <param name="operation">The async operation to protect.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>A <see cref="D2Result{TValue}"/> wrapping the outcome.</returns>
    public async ValueTask<D2Result<TValue>> ExecuteAsync(
        TKey key,
        Func<CancellationToken, ValueTask<TValue>> operation,
        CancellationToken ct = default)
    {
        try
        {
            // Walk inner-to-outer, wrapping the operation with each layer.
            // Each iteration captures its own copies of `layer` and `inner`
            // so closure semantics are correct.
            var wrapped = operation;
            for (var i = r_layersOuterFirst.Length - 1; i >= 0; i--)
            {
                var layer = r_layersOuterFirst[i];
                var inner = wrapped;
                wrapped = c => layer.WrapAsync(key, inner, c);
            }

            return D2Result<TValue>.Ok(await wrapped(ct));
        }
        catch (CircuitOpenException)
        {
            return D2Result<TValue>.ServiceUnavailable();
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested)
        {
            return D2Result<TValue>.Cancelled();
        }
        catch (Exception ex) when (RetryHelper.IsTransientException(ex))
        {
            return D2Result<TValue>.ServiceUnavailable();
        }
        catch
        {
            return D2Result<TValue>.UnhandledException();
        }
    }
}

File: server/shared/typescript/resilience/src/pipeline/resilient-pipeline.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import type { CircuitBreakerOptions } from "../circuit-breaker/circuit-breaker-options.js";
import type { RetryOptions } from "../retry/retry-options.js";
import { CircuitBreakerLayer } from "./circuit-breaker-layer.js";
import type { IResilientLayer } from "./i-resilient-layer.js";
import type { RateLimiterOptions } from "./rate-limiter-layer.js";
import { RateLimiterLayer } from "./rate-limiter-layer.js";
import { RetryLayer } from "./retry-layer.js";
import { SingleflightLayer } from "./singleflight-layer.js";
import type { TimeoutOptions } from "./timeout-layer.js";
import { TimeoutLayer } from "./timeout-layer.js";

/**
 * Final composed pipeline. Outer-first ordering.
 *
 * Use {@link ResilientPipelineBuilder} to compose layers. The canonical
 * full-stack order (outermost → innermost) mirrors the .NET standard:
 * ```
 * Singleflight → RateLimiter → TotalTimeout → Retry → CircuitBreaker → PerAttemptTimeout
 * ```
 *
 * **Order matters.** `CircuitBreaker` outside `Retry` trips after N total
 * failures across all attempts; `CircuitBreaker` inside `Retry` trips after
 * one burst of N consecutive per-attempt failures (re-arms on every outer
 * retry). Use `PassThrough` for a zero-layer pipeline that still executes the
 * op (suitable as a bypass sentinel).
 */
export class ResilientPipeline {
  constructor(private readonly head: IResilientLayer) {}

  /**
   * Executes `op` through the configured layer stack. The optional `signal`
   * (≈ .NET `CancellationToken`) threads down every layer; the op receives the
   * signal it should observe for cooperative cancellation (a {@link TimeoutLayer}
   * substitutes a signal that ALSO fires on its deadline).
   */
  execute<T>(
    key: string,
    op: (signal?: AbortSignal) => Promise<T>,
    signal?: AbortSignal,
  ): Promise<T> {
    return this.head.execute(key, op, signal);
  }

  /**
   * A zero-layer pass-through pipeline. Runs the op directly with no
   * resilience wrapping. Mirrors .NET `ResilientPipeline.PassThrough`.
   *
   * Use as an explicit bypass: the caller opts for a raw call while retaining
   * the same `ResilientPipeline` call-site interface.
   */
  static get PassThrough(): ResilientPipeline {
    return new ResilientPipeline(new TerminalLayer());
  }
}

/** Innermost layer — runs the supplied op directly with the threaded signal. */
class TerminalLayer implements IResilientLayer {
  execute<T>(
    _key: string,
    op: (signal?: AbortSignal) => Promise<T>,
    signal?: AbortSignal,
  ): Promise<T> {
    return op(signal);
  }
}

/**
 * Builder for {@link ResilientPipeline}. Outer-first ordering — the
 * first layer added is the outermost wrapper. Mirrors .NET
 * `ResilientPipelineBuilder`.
 *
 * Canonical full-stack composition (add in this order):
 * ```
 * new ResilientPipelineBuilder()
 *   .useSingleflight()                                   // outermost (optional)
 *   .useRateLimiter({ maxConcurrency: 10 })              // admission control
 *   .useTimeout({ durationMs: 30_000 })                  // total-request budget
 *   .useRetries({ maxAttempts: 3 })
 *   .useCircuitBreaker({ failureThreshold: 5, cooldownMs: 30_000 })
 *   .useTimeout({ durationMs: 5_000 })                   // per-attempt deadline
 *   .build()
 * ```
 */
export class ResilientPipelineBuilder {
  private readonly layers: ((inner: IResilientLayer) => IResilientLayer)[] = [];

  useSingleflight(): this {
    this.layers.push((inner) => new SingleflightLayer(inner));
    return this;
  }

  useCircuitBreaker(opts: CircuitBreakerOptions): this {
    this.layers.push((inner) => new CircuitBreakerLayer(inner, opts));
    return this;
  }

  useRetries(opts: Partial<RetryOptions<unknown>>): this {
    this.layers.push((inner) => new RetryLayer(inner, opts));
    return this;
  }

  /**
   * Adds a {@link TimeoutLayer} at the current position. Call twice to express
   * both a total-request timeout (outer) and a per-attempt timeout (inner,
   * below `useRetries`).
   *
   * `opts.durationMs <= 0` disables the timeout (pass-through for this position).
   */
  useTimeout(opts?: TimeoutOptions): this {
    this.layers.push((inner) => new TimeoutLayer(inner, opts));
    return this;
  }

  /**
   * Adds a {@link RateLimiterLayer} at the current position. Place outermost
   * (after `useSingleflight` if used) for admission-control semantics — reject
   * callers before they consume retry or timeout budget.
   *
   * `opts.maxConcurrency` must be >= 1 (validated at construction).
   */
  useRateLimiter(opts?: RateLimiterOptions): this {
    this.layers.push((inner) => new RateLimiterLayer(inner, opts));
    return this;
  }

  build(): ResilientPipeline {
    let head: IResilientLayer = new TerminalLayer();
    // Build innermost-out so order matches insertion (first added = outermost).
    for (let i = this.layers.length - 1; i >= 0; i--)
      head = this.layers[i]!(head);
    return new ResilientPipeline(head);
  }
}
